gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaAvroSchemaRegistry.java

```diff
@@ -0,0 +1,289 @@
+
+  private static final int GET_SCHEMA_BY_ID_MAX_TIRES = 3;
+  private static final int GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS = 1;
+  private static final String GET_RESOURCE_BY_ID = "/id=";
+  private static final String GET_RESOURCE_BY_TYPE = "/latest_with_type=";
+  private static final String SCHEMA_ID_HEADER_NAME = "Location";
+  private static final String SCHEMA_ID_HEADER_PREFIX = "/id=";
+
+  private final LoadingCache<String, Schema> cachedSchemasById;
+
+  private final HttpClient httpClient;
+
+  private final String url;
+
+  public KafkaAvroSchemaRegistry(String url) {
+    this.url = url;
+    this.cachedSchemasById =
+        CacheBuilder.newBuilder().maximumSize(1000).expireAfterWrite(10, TimeUnit.MINUTES)
+            .build(new KafkaSchemaCacheLoader());
+    this.httpClient = new HttpClient(new MultiThreadedHttpConnectionManager());
+  }
+
+  /**
+   * Get schema from schema registry by ID
+   *
+   * @param id Schema ID
+   * @return Schema with the corresponding ID
+   * @throws SchemaNotFoundException if ID not found
+   */
+  public synchronized Schema getSchemaById(String id) throws SchemaNotFoundException {
+    try {
+      return cachedSchemasById.get(id);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Get the latest schema of a topic.
+   *
+   * @param topic topic name
+   * @return the latest schema
+   * @throws SchemaNotFoundException if topic name not found.
+   */
+  public synchronized Schema getLatestSchemaByTopic(String topic) throws SchemaNotFoundException {
+    String schemaUrl = KafkaAvroSchemaRegistry.this.url + GET_RESOURCE_BY_TYPE + topic;
+
+    LOG.debug("Fetching from URL : " + schemaUrl);
+
+    GetMethod get = new GetMethod(schemaUrl);
+
+    int statusCode;
+    String schemaString;
+    try {
+      statusCode = KafkaAvroSchemaRegistry.this.httpClient.executeMethod(get);
+      schemaString = get.getResponseBodyAsString();
+    } catch (HttpException e) {
+      throw new RuntimeException(e);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } finally {
+      get.releaseConnection();
+    }
+
+    if (statusCode != HttpStatus.SC_OK) {
+      throw new SchemaNotFoundException(String.format("Latest schema for topic %s cannot be retrieved", topic));
+    }
+
+    Schema schema;
+    if (schemaString.startsWith("{")) {
+      try {
+        schema = new Schema.Parser().parse(schemaString);
+      } catch (Exception e) {
+        throw new SchemaNotFoundException(String.format("Latest schema for topic %s cannot be retrieved", topic));
+      }
+    } else {
+      throw new SchemaNotFoundException(String.format("Latest schema for topic %s cannot be retrieved", topic));
+    }
+
+    return schema;
+  }
+
+  /**
+   * Register a schema to the Kafka schema registry
+   *
+   * @param schema
+   * @return schema ID of the registered schema
+   */
+  public synchronized String register(Schema schema) {
```

Comment:
We need to justify the current usage based on what kind of guarantee we need here. This only prevents more than one thread in the same JVM from updating the schema registry. Again, it's good to know what kind of guarantees we need here first.